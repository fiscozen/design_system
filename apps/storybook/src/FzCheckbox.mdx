import { Meta } from '@storybook/addon-docs/blocks'

<Meta title="Documentation/Form/Checkbox" />

# FzCheckbox

A fully accessible checkbox component with support for single selection (boolean), multi-selection (array), indeterminate states, and hierarchical relationships.

## Features

- Single checkbox (boolean v-model)
- Indeterminate state for parent checkboxes
- Emphasis mode for highlighted checkboxes
- Error state with validation messages
- Disabled state
- Tooltip support
- Full keyboard navigation
- WCAG 2.1 AA compliant

## Installation

```bash
pnpm add @fiscozen/checkbox
```

## Props

<table>
  <thead>
    <tr>
      <th>Prop</th>
      <th>Type</th>
      <th>Default</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>label</code></td>
      <td><code>string</code></td>
      <td>-</td>
      <td>Text label displayed next to the checkbox</td>
    </tr>
    <tr>
      <td><code>value</code></td>
      <td><code>string | number | boolean</code></td>
      <td>-</td>
      <td>Value used in array mode (multi-select)</td>
    </tr>
    <tr>
      <td><code>indeterminate</code></td>
      <td><code>boolean</code></td>
      <td><code>false</code></td>
      <td>Shows indeterminate state (partial selection)</td>
    </tr>
    <tr>
      <td><code>emphasis</code></td>
      <td><code>boolean</code></td>
      <td><code>false</code></td>
      <td>Applies emphasis styling (blue color)</td>
    </tr>
    <tr>
      <td><code>error</code></td>
      <td><code>boolean</code></td>
      <td><code>false</code></td>
      <td>Shows error state with red color</td>
    </tr>
    <tr>
      <td><code>disabled</code></td>
      <td><code>boolean</code></td>
      <td><code>false</code></td>
      <td>Disables the checkbox</td>
    </tr>
    <tr>
      <td><code>required</code></td>
      <td><code>boolean</code></td>
      <td><code>false</code></td>
      <td>Marks the checkbox as required</td>
    </tr>
    <tr>
      <td><code>standalone</code></td>
      <td><code>boolean</code></td>
      <td><code>false</code></td>
      <td>Hides the label (checkbox only)</td>
    </tr>
    <tr>
      <td><code>tooltip</code></td>
      <td><code>object</code></td>
      <td>-</td>
      <td>Tooltip configuration object</td>
    </tr>
    <tr>
      <td><code>checkboxId</code></td>
      <td><code>string</code></td>
      <td>auto-generated</td>
      <td>Custom ID for the checkbox input</td>
    </tr>
    <tr>
      <td><code>ariaOwns</code></td>
      <td><code>string</code></td>
      <td>-</td>
      <td>ARIA owns attribute for parent-child relationships</td>
    </tr>
  </tbody>
</table>

## Slots

<table>
  <thead>
    <tr>
      <th>Slot</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>error</code></td>
      <td>Error message content (shown when <code>error={true}</code>)</td>
    </tr>
    <tr>
      <td><code>children</code></td>
      <td>Child checkboxes for hierarchical structures</td>
    </tr>
  </tbody>
</table>

## Usage Examples

### Single Checkbox (Boolean)

```vue
<template>
  <FzCheckbox v-model="accepted" label="I accept the terms and conditions" />
</template>

<script setup>
import { ref } from 'vue'
import { FzCheckbox } from '@fiscozen/checkbox'

const accepted = ref(false)
</script>
```

### Emphasis Mode

```vue
<FzCheckbox v-model="selected" label="Important option" emphasis />
```

### Error State

```vue
<FzCheckbox v-model="selected" label="Required field" error>
  <template #error>
    This field is required
  </template>
</FzCheckbox>
```

### Indeterminate State

```vue
<FzCheckbox v-model="selected" label="Select all" :indeterminate="true" />
```

### With Tooltip

```vue
<FzCheckbox 
  v-model="selected" 
  label="Advanced option"
  :tooltip="{ content: 'This option enables advanced features' }"
/>
```

### Disabled

```vue
<FzCheckbox v-model="selected" label="Disabled option" disabled />
```

## Accessibility

- **Keyboard navigation**: Full support for Tab, Space, and Enter keys
- **Screen readers**: Proper ARIA labels and state announcements
- **Focus indicators**: Visible focus ring for keyboard navigation
- **Semantic HTML**: Uses native checkbox input for proper form behavior
- **WCAG 2.1 AA**: Meets color contrast and accessibility requirements

### ARIA Attributes

- `aria-checked`: Announces checked/unchecked/mixed (indeterminate) state
- `aria-label`: Provides accessible name when standalone
- `aria-labelledby`: Links to visible label text
- `aria-describedby`: Links to error messages and help text
- `aria-required`: Indicates required fields
- `aria-invalid`: Indicates validation errors
- `aria-owns`: Establishes parent-child relationships

## Notes

### Deprecated Props

- **`size`**: The size prop is deprecated and will be removed in a future version. Checkboxes now have a fixed size (md).
