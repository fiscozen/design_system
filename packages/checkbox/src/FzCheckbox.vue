<script setup lang="ts">
/**
 * FzCheckbox Component
 *
 * A fully accessible checkbox component with support for single selection (boolean),
 * multi-selection (array), indeterminate states, and hierarchical relationships.
 *
 * @component
 * @example
 * // Single checkbox (boolean)
 * <FzCheckbox v-model="accepted" label="I accept terms" />
 *
 * @example
 * // Multi-select checkbox (array)
 * <FzCheckbox v-model="selectedItems" value="item1" label="Item 1" />
 *
 * @example
 * // Indeterminate checkbox (parent with children)
 * <FzCheckbox v-model="selection" :indeterminate="true" label="Select All" />
 */
import { computed, onMounted, shallowRef } from "vue";
import { FzCheckboxProps } from "./types";
import {
  mapSizeToClasses,
  CHECKBOX_ICONS,
  CHECKBOX_ICON_VARIANTS,
} from "./common";
import { generateCheckboxId } from "./utils";
import { FzIcon } from "@fiscozen/icons";
import { FzTooltip } from "@fiscozen/tooltip";
import { FzAlert } from "@fiscozen/alert";

const props = withDefaults(defineProps<FzCheckboxProps>(), {
  size: "md",
  indeterminate: false,
});

/**
 * Computes the actual value to use for the checkbox.
 * Falls back to label if no explicit value is provided.
 */
const currentValue = computed(() => props.value ?? props.label);

/**
 * Unique identifier for the checkbox input.
 * Uses provided checkboxId prop if available, otherwise generates unique ID
 * using timestamp + random suffix strategy.
 * Deterministic IDs are important for aria-owns relationships in hierarchical structures.
 */
const id = props.checkboxId || generateCheckboxId();

/**
 * Two-way binding for checkbox state.
 * Supports three modes:
 * - null/undefined: Uncontrolled state
 * - boolean: Single checkbox (true/false)
 * - array: Multi-select checkbox group (contains values of checked items)
 */
const model = defineModel<
  null | undefined | boolean | (string | number | boolean)[]
>({
  required: true,
});

const emit = defineEmits(["change"]);

/**
 * Reference to the native checkbox input element.
 * Uses shallowRef for optimal performance with DOM element references.
 * ShallowRef avoids deep reactivity tracking of all DOM element properties.
 *
 * @see https://vuejs.org/api/reactivity-advanced.html#shallowref
 * @note When upgrading to Vue 3.5+, consider migrating to useTemplateRef()
 */
const refCheckbox = shallowRef<HTMLInputElement | null>(null);

/**
 * CSS classes for the hidden native checkbox input.
 * Uses "peer" for Tailwind's peer selector to style adjacent elements based on input state.
 * Input is visually hidden but remains accessible to screen readers and keyboard navigation.
 */
const staticInputClass = "w-0 h-0 peer fz-hidden-input";

/**
 * CSS classes for the label element.
 * Includes focus ring styles that appear on the icon when the hidden input receives focus.
 * The pseudo-element (after:) creates a visible focus indicator for keyboard navigation.
 */
const staticLabelClass = `
  flex gap-4 items-center hover:cursor-pointer text-core-black
  peer-focus:[&_div]:after:border-1
  peer-focus:[&_div]:after:border-solid
  peer-focus:[&_div]:after:rounded-[3px]
  peer-focus:[&_div]:after:border-blue-500
  peer-focus:[&_div]:after:content-['']
  peer-focus:[&_div]:after:top-0
  peer-focus:[&_div]:after:left-0
  peer-focus:[&_div]:after:right-0
  peer-focus:[&_div]:after:bottom-0
  peer-focus:[&_div]:after:absolute
`;

/** Position context for the focus ring pseudo-element */
const staticIconClass = "relative";

/**
 * Determines if the checkbox is currently checked.
 * Handles both boolean and array models:
 * - Boolean model: Returns the boolean value directly
 * - Array model: Returns true if the array contains this checkbox's value
 * - Null/undefined: Returns false
 */
const isChecked = computed(() => {
  if (model.value == null) return false;

  if (typeof model.value === "boolean") {
    return model.value;
  } else {
    return model.value.includes(currentValue.value);
  }
});

/** Dynamic label classes based on size and state (disabled, error, emphasis) */
const computedLabelClass = computed(() => [
  mapSizeToClasses[props.size],
  getTextClassForLabel(),
]);

/** Dynamic icon classes based on size and state */
const computedIconClasses = computed(() => [
  props.size === "sm" ? "mt-1" : "",
  getTextClassForIcon(),
]);

/**
 * Determines which FontAwesome icon to display:
 * - Indeterminate: square with minus (partial selection)
 * - Checked: square with check
 * - Unchecked: empty square
 */
const computedName = computed(() => {
  if (props.indeterminate) {
    return CHECKBOX_ICONS.INDETERMINATE;
  }

  return isChecked.value ? CHECKBOX_ICONS.CHECKED : CHECKBOX_ICONS.UNCHECKED;
});

/**
 * FontAwesome icon variant:
 * - "fas" (solid): For checked or indeterminate states
 * - "far" (regular): For unchecked state
 */
const computedVariant = computed(() => {
  return props.indeterminate || isChecked.value
    ? CHECKBOX_ICON_VARIANTS.SOLID
    : CHECKBOX_ICON_VARIANTS.REGULAR;
});

/**
 * Computes text color classes for the label based on checkbox state.
 * Priority order: disabled > error > emphasis > default
 *
 * @returns Tailwind CSS classes for label text color
 */
const getTextClassForLabel = () => {
  switch (true) {
    case props.disabled:
      return "text-grey-300";
    case props.error:
      return "text-semantic-error";
    case props.emphasis:
      // Emphasis mode: icon changes to blue when checked/indeterminate
      return "text-core-black peer-checked:[&_div]:text-blue-500 peer-indeterminate:[&_div]:text-blue-500";
    default:
      return "text-core-black";
  }
};

/**
 * Computes text color classes for the checkbox icon.
 * Priority order: disabled > error > default
 * Note: Emphasis state falls through to default (grey) for unchecked state
 *
 * @returns Tailwind CSS classes for icon color
 */
const getTextClassForIcon = () => {
  switch (true) {
    case props.disabled:
      return "text-grey-300";
    case props.error:
      return "text-semantic-error";
    case props.emphasis:
    default:
      return "text-grey-500";
  }
};

/**
 * Lifecycle hook: Triggers change event on mount if checkbox is initially checked.
 * This ensures that any parent components listening to change events receive
 * initial state, which is particularly important for parent checkboxes that need
 * to synchronize with their children's initial state.
 */
onMounted(() => {
  if (model.value == null) return;

  // Boolean model: dispatch change if true
  if (typeof model.value === "boolean") {
    if (model.value) refCheckbox.value?.dispatchEvent(new Event("change"));
  }
  // Array model: dispatch change if this checkbox's value is in the array
  else {
    if (model.value.includes(currentValue.value))
      refCheckbox.value?.dispatchEvent(new Event("change"));
  }
});
</script>

<template>
  <!-- Root container: vertical layout with consistent spacing -->
  <div class="flex justify-center flex-col w-fit gap-4">
    <!-- Checkbox row: input + label + optional tooltip -->
    <div class="flex gap-0 items-center">
      <!-- 
        Native checkbox input (visually hidden but functionally present)
        - Maintains native keyboard navigation and form behavior
        - Screen readers can interact with it normally
        - Styled via adjacent label and icon using Tailwind's "peer" selector
      -->
      <input
        type="checkbox"
        :id="id"
        :disabled="disabled"
        :class="staticInputClass"
        :required="required"
        :value="value"
        @change="emit('change', $event)"
        v-model="model"
        :aria-checked="indeterminate ? 'mixed' : isChecked"
        :aria-label="label"
        :aria-required="required"
        :aria-invalid="error"
        :aria-describedby="error ? `${id}-error` : undefined"
        :aria-labelledby="standalone ? undefined : `${id}-label`"
        :aria-owns="props.ariaOwns"
        ref="refCheckbox"
      />
      <!-- 
        Label element: provides click target and visual representation
        Connected to hidden input via "for" attribute
      -->
      <label
        :id="`${id}-label`"
        :for="id"
        :class="[staticLabelClass, computedLabelClass]"
      >
        <!-- 
          Visual checkbox icon (replaces hidden native checkbox)
          aria-hidden because it's purely decorative - the native input conveys state
          @TODO: When FzIcon natively supports ariaHidden prop, remove aria-hidden as HTML attribute and use :aria-hidden as prop
        -->
        <FzIcon
          :name="computedName"
          :size="size"
          :class="[staticIconClass, computedIconClasses]"
          :variant="computedVariant"
          aria-hidden="true"
        />
        <!-- Label text (hidden in standalone mode for checkbox-only display) -->
        <template v-if="!standalone">{{ label }}</template>
      </label>

      <!-- 
        Optional tooltip for additional context
        @TODO: When FzTooltip supports keyboard accessibility and ARIA attributes (role="tooltip", aria-describedby), update implementation
      -->
      <FzTooltip v-if="tooltip" v-bind="tooltip" class="ml-4">
        <!-- 
          Info icon (informative, not decorative)
          Has role="img" and aria-label because it conveys meaning
          @TODO: When FzIcon natively supports role and ariaLabel props, use them instead of HTML attributes
        -->
        <FzIcon
          name="info-circle"
          :size="size"
          class="text-semantic-info"
          role="img"
          aria-label="Informazioni aggiuntive"
        />
      </FzTooltip>
    </div>
    <!-- 
      Error message display
      Uses ARIA live region to announce errors to screen readers immediately
      - role="alert": Identifies as important message
      - aria-live="assertive": Interrupts current screen reader announcement
      - aria-atomic="true": Reads entire message, not just changes
      @TODO: When FzAlert natively supports role, ariaLive, ariaAtomic as props, remove HTML attributes and use typed props
    -->
    <FzAlert
      v-if="error && $slots.error"
      :id="`${id}-error`"
      :size="size"
      type="error"
      alertStyle="simple"
      role="alert"
      aria-live="assertive"
      aria-atomic="true"
    >
      <slot name="error" />
    </FzAlert>

    <!-- 
      Children slot for nested checkboxes
      Used by FzCheckboxGroupOption to render child checkboxes
    -->
    <slot name="children" />
  </div>
</template>

<style scoped>
/**
 * Visually hides the native checkbox input while keeping it accessible.
 * 
 * Why not display:none or visibility:hidden?
 * - Those would remove the input from the accessibility tree
 * - Screen readers couldn't interact with it
 * - Keyboard navigation would be broken
 * 
 * This approach:
 * - Maintains accessibility tree presence
 * - Preserves keyboard focus capability
 * - Allows screen reader interaction
 * - Enables form submission behavior
 */
.fz-hidden-input {
  opacity: 0;
  margin: 0;
  height: 0;
  border: 0 none;
  appearance: none;
}
</style>
