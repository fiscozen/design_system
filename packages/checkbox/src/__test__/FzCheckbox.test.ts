// FzCheckbox.spec.ts
import { mount } from "@vue/test-utils";
import { describe, it, expect, beforeEach } from "vitest";
import FzCheckbox from "../FzCheckbox.vue";

describe("FzCheckbox", () => {
  it("renders correctly", async () => {
    const wrapper = mount(FzCheckbox, {
      props: {
        label: "Test Checkbox",
        value: "test",
        size: "md",
        modelValue: false,
      },
    });

    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toContain("Test Checkbox");
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("emits an update event when clicked", async () => {
    const wrapper = mount(FzCheckbox, {
      props: {
        label: "Test Checkbox",
        value: "test",
        modelValue: false,
      },
    });
    await wrapper.find("input").trigger("click");
    expect(wrapper.emitted()).toHaveProperty("update:modelValue");
  });

  it("is checked when v-model is true", async () => {
    const wrapper = mount(FzCheckbox, {
      props: {
        label: "Test Checkbox",
        value: "test",
        modelValue: true,
      },
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.find("input").element.checked).toBe(true);
  });

  it("is unchecked when v-model is false", async () => {
    const wrapper = mount(FzCheckbox, {
      props: {
        label: "Test Checkbox",
        value: "test",
        modelValue: false,
      },
    });

    await wrapper.vm.$nextTick();
    expect(wrapper.find("input").element.checked).toBe(false);
  });

  it("should be emphasized", async () => {
    const wrapper = mount(FzCheckbox, {
      props: {
        label: "Test Checkbox",
        value: "test",
        size: "md",
        modelValue: false,
        emphasis: true,
      },
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.find("label").classes()).toContain(
      "peer-checked:[&_div]:text-blue-500",
    );
  });

  it("should be disabled", async () => {
    const wrapper = mount(FzCheckbox, {
      props: {
        label: "Test Checkbox",
        value: "test",
        size: "md",
        modelValue: false,
        disabled: true,
      },
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.find("input").element.disabled).toBe(true);
  });
});
