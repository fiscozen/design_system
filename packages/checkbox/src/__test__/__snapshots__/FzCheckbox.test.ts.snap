// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FzCheckbox > renders correctly 1`] = `
"<div data-v-79ecb611="" class="flex justify-center flex-col w-fit"><input data-v-79ecb611="" type="checkbox" id="test" class="checkbox--medium peer" value="test"><label data-v-79ecb611="" for="test" class="flex items-start gap-4 text-md text-grey-500">
    <div data-v-79ecb611="" class="flex items-center justify-center w-[20px] h-[20px] mt-2 text-grey-500 relative"><svg class="svg-inline--fa fa-square h-[16px]" aria-hidden="true" focusable="false" data-prefix="far" data-icon="square" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512">
        <path class="" fill="currentColor" d="M384 80c8.8 0 16 7.2 16 16V416c0 8.8-7.2 16-16 16H64c-8.8 0-16-7.2-16-16V96c0-8.8 7.2-16 16-16H384zM64 32C28.7 32 0 60.7 0 96V416c0 35.3 28.7 64 64 64H384c35.3 0 64-28.7 64-64V96c0-35.3-28.7-64-64-64H64z"></path>
      </svg></div><span data-v-79ecb611="" class="w-fit">Test Checkbox</span>
  </label>
  <!--v-if-->
</div>"
`;
